clients/codelist/codelist.go

```diff
@@ -124,107 +130,165 @@ func (c *Client) GetIDNameMap(id string) (map[string]string, error) {
 	if err != nil {
-		return
+		return codes, err
+	}
+
+	defer closeResponseBody(ctx, resp)
+
+	if resp.StatusCode != http.StatusOK {
+		return codes, &ErrInvalidCodelistAPIResponse{http.StatusOK, resp.StatusCode, url}
 	}
-	defer resp.Body.Close()
 
 	b, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return
+		return codes, err
 	}
 
 	err = json.Unmarshal(b, &codes)
 	if err != nil {
-		return
+		return codes, err
 	}
+
 	return codes, nil
 }
 
-// GetCodeByID returns informtion about a code
-func (c *Client) GetCodeByID(codeListID string, edition string, codeID string) (code CodeResult, err error) {
+// GetCodeByID returns information about a code
+func (c *Client) GetCodeByID(ctx context.Context, serviceAuthToken string, codeListID string, edition string, codeID string) (CodeResult, error) {
+	var code CodeResult
 	url := fmt.Sprintf("%s/code-lists/%s/editions/%s/codes/%s", c.url, codeListID, edition, codeID)
-	resp, err := c.cli.Get(context.Background(), url)
+
+	resp, err := c.doServiceRequest(ctx, serviceAuthToken, "GET", url, nil)
 	if err != nil {
-		return
+		return code, err
 	}
 
-	defer resp.Body.Close()
+	defer closeResponseBody(ctx, resp)
+
+	if resp.StatusCode != http.StatusOK {
+		return code, &ErrInvalidCodelistAPIResponse{http.StatusOK, resp.StatusCode, url}
+	}
 
 	b, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return
+		return code, err
 	}
 
 	err = json.Unmarshal(b, &code)
 	if err != nil {
-		return
+		return code, err
 	}
+
 	return code, nil
 }
 
-func (c *Client) GetDatasetsByCode(codeListID string, edition string, codeID string) (datasets DatasetsResult, err error) {
+// GetDatasetsByCode todo
+func (c *Client) GetDatasetsByCode(ctx context.Context, serviceAuthToken string, codeListID string, edition string, codeID string) (DatasetsResult, error) {
+	var datasets DatasetsResult
 	url := fmt.Sprintf("%s/code-lists/%s/editions/%s/codes/%s/datasets", c.url, codeListID, edition, codeID)
-	resp, err := c.cli.Get(context.Background(), url)
+
+	resp, err := c.doServiceRequest(ctx, serviceAuthToken, "GET", url, nil)
 	if err != nil {
-		return
+		return datasets, err
 	}
 
-	defer resp.Body.Close()
+	defer closeResponseBody(ctx, resp)
+
+	if resp.StatusCode != http.StatusOK {
+		return datasets, &ErrInvalidCodelistAPIResponse{http.StatusOK, resp.StatusCode, url}
+	}
 
 	b, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return
+		return datasets, err
 	}
 
 	err = json.Unmarshal(b, &datasets)
 	if err != nil {
-		return
+		return datasets, err
 	}
 	return datasets, nil
 }
+
+// doServiceRequest executes clienter.Do setting the service authentication token as a request header. Returns the http.Response and any error.
+// It is the callers responsibility to ensure response.Body is closed on completion.
+func (c *Client) doServiceRequest(ctx context.Context, serviceAuthToken string, method string, uri string, body io.Reader) (*http.Response, error) {
```

Comment:
I really like the use of these helper functions especially the `closeResponseBody` I think the other APIs would benefit from these too. instead of us adding these functions to each go-ns/client package it might be worth us creating some utilities as part of the dp-rchttp package which the other APIs can then use?

Having said that the doServiceRequest to me sounds like a function that just 'sends a service request' as per the name. However, technically the function is manipulating the request and not just sending it, so the function is doing more than it says it is in the name (although function annotation does state it will do this).

Reply:
I would counter argue that the request is not exposed to the caller - it is scoped within the function so its valid for the function do manipulate it however it needs to.

But I have renamed it to `doRequestWithServiceAuthHeader` for further transparency.